here-oauth-client/src/main/java/com/here/account/http/HttpProvider.java

```
@@ -113,13 +114,14 @@ void authorize(HttpRequest httpRequest, String method, String url,
         InputStream getResponseBody() throws IOException;
 
         /**
-         * Returns all the headers from the response
-         * @return returns a Map of headers if the method implementation returns the headers 
-         *      or throws Unsupported Operation Exception if the method is not implemented
+         * Returns all the headers from the response.
+         *
+         * @return returns a Map of headers,
+         *      or an empty map if the implementing class has not overridden the default behavior.
          */
 
         default Map<String, List<String>> getHeaders() {
-            throw new UnsupportedOperationException();
+            return Collections.emptyMap();
```

This is a misleading default implementation for this method.  Any implementations of HttpResponse that do not override this method will now return an empty collection which does not necessarily represent the actual headers of the response. Throwing an exception in these cases is much more appropriate.

I see that this method is now being used in the core implementation of the Client but not accounting for the possibility of the default behavior.  Perhaps the use of this method (which looks for the Content-Type of the response) in the Client should be refactored to "assume JSON" when getHeaders throws an UnsupportedOperationException rather than providing this default implementation?

---

This was my thought, too.  I don't disagree with your arguments, I just didn't like the ugly code that results in the Client class.  If I make this change, then with the ugly UnsupportedOperationException-handling code in Client will be a necessary evil until we can move on to a non-backward-compatible major version change.